internal/crypto/cms/encoding/ber/ber.go

```diff
@@ -0,0 +1,301 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ber decodes BER-encoded ASN.1 data structures and encodes in DER.
+// Note:
+//   - DER is a subset of BER.
+//   - Indefinite length is not supported.
+//   - The length of the encoded data must fit the memory space of the int type
+//     (4 bytes).
+//
+// Reference:
+// - http://luca.ntop.org/Teaching/Appunti/asn1.html
+// - ISO/IEC 8825-1:2021
+package ber
+
+import (
+	"bytes"
+	"encoding/asn1"
+	"fmt"
+)
+
+// value is the interface for an ASN.1 value node.
+type value interface {
+	// EncodeMetadata encodes the identifier and length in DER to the buffer.
+	EncodeMetadata(*bytes.Buffer) error
+
+	// EncodedLen returns the length in bytes of the encoded data.
+	EncodedLen() int
+
+	// Content returns the content of the value.
+	// For primitive values, it returns the content octets.
+	// For constructed values, it returns nil because the content is
+	// the data of all members.
+	Content() []byte
+}
+
+// ConvertBERToDER converts BER-encoded ASN.1 data structures to DER-encoded.
+func ConvertBERToDER(ber []byte) ([]byte, error) {
+	if ber == nil || len(ber) == 0 {
+		return nil, asn1.SyntaxError{Msg: "BER-encoded ASN.1 data structures is empty"}
+	}
+
+	flatValues, err := decode(ber)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the total length from the root value and allocate a buffer
+	buf := bytes.NewBuffer(make([]byte, 0, flatValues[0].EncodedLen()))
+	for _, v := range flatValues {
+		if err = v.EncodeMetadata(buf); err != nil {
+			return nil, err
+		}
+
+		if content := v.Content(); content != nil {
+			// primitive value
+			_, err = buf.Write(content)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+// decode decodes BER-encoded ASN.1 data structures.
+// To get the DER of `r`, encode the values
+// in the returned slice in order.
+//
+// Parameters:
+// r - The input byte slice.
+//
+// Return:
+// []value - The returned value, which is the flat slice of ASN.1 values,
+// contains the nodes from a depth-first traversal.
+// error - An error that can occur during the decoding process.
+//
+// Reference: ISO/IEC 8825-1: 8.1.1.3
+func decode(r []byte) ([]value, error) {
+	// prepare the first value
+	identifier, contentLen, r, err := decodeMetadata(r)
+	if err != nil {
+		return nil, err
+	}
+	if len(r) != contentLen {
+		return nil, asn1.StructuralError{Msg: fmt.Sprintf("length octets value %d does not match with content length %d", contentLen, len(r))}
+	}
+
+	// primitive value
+	if isPrimitive(identifier) {
+		return []value{&primitive{
+			identifier: identifier,
+			content:    r,
+		}}, nil
+	}
+
+	// constructed value
+	rootConstructed := &constructed{
+		identifier: identifier,
+		rawContent: r,
+	}
+	flatValues := []value{rootConstructed}
+
+	// start depth-first decoding with stack
+	valueStack := []*constructed{rootConstructed}
```

Comment:
`valueStack` is essentially a stack of `constructed` values only. Let's rename it:
```suggestion
	contructedStack := []*constructed{rootConstructed}
```

Reply:
Updated